### PR TITLE
Address additional feedback from number-handling PRs

### PR DESCRIPTION
**Pull request by layomia**
Follow up for https://github.com/dotnet/runtime/pull/39363 & https://github.com/dotnet/runtime/pull/39716. 

FYI @ahsonkhan

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj [hunk 1]
```diff
@@ -227,11 +226,9 @@
   <ItemGroup Condition="$(TargetFramework.StartsWith('netstandard')) or $(TargetFramework.StartsWith('net4'))">
     <Compile Include="System\Collections\Generic\StackExtensions.netstandard.cs" />
     <!-- Common or Common-branched source files -->
-    <Compile Include="$(CommonPath)System\Buffers\ArrayBufferWriter.cs"
-             Link="Common\System\Buffers\ArrayBufferWriter.cs" />
+    <Compile Include="$(CommonPath)System\Buffers\ArrayBufferWriter.cs" Link="Common\System\Buffers\ArrayBufferWriter.cs" />
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)' or
-                        '$(TargetFramework)' == 'netcoreapp3.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)' or&#xD;&#xA;                        '$(TargetFramework)' == 'netcoreapp3.0'">
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System.Text.Json.csproj:231]: Revert unrelated changes. Also, I am curious to know where `&#xD;` came from (which tool/editor)?

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs [hunk 2]
```diff
@@ -134,5 +135,10 @@ public static bool IsFinite(float value)
             return !(float.IsNaN(value) || float.IsInfinity(value));
 #endif
         }
+
+        public static bool IsValidNumberHandlingValue(JsonNumberHandling handling)
+        {
+            return JsonHelpers.IsInRangeInclusive((int)handling, 0, 7);
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs:141]: nit: Do this instead, to avoid magic numbers
```suggestion
            return JsonHelpers.IsInRangeInclusive((int)handling, 0, (int)(Strict |  AllowReadingFromString |  WriteAsString | AllowNamedFloatingPointLiterals));
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs [hunk 3]
```diff
@@ -29,14 +29,15 @@ internal abstract class DictionaryDefaultConverter<TCollection, TKey, TValue>
...
 
         internal override Type ElementType => s_valueType;
 
-        protected Type KeyType = typeof(TKey);
+        protected static readonly Type KeyType = typeof(TKey);
+
         // For string keys we don't use a key converter
         // in order to avoid performance regression on already supported types.
-        protected bool IsStringKey = typeof(TKey) == typeof(string);
+        private static readonly bool s_isStringKey = KeyType == typeof(string);
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs:40]: Generally speaking, when the `(x == typeof(...))` branch is false, that whole code path is optimized out. Will  that remain true with this pattern as well, along with how we use `ElementType`?

For example:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L434-L455

[sharplab.io example](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4CS51ATgPYAOAyswG4UDG1CAG50WXARLkqtAMIsobCgBtmXJrwHDRODgAswTNgBkwAIyJlKNEWlEBmcQDZcCbDOwBvdNh/ZvvgG0AKQpgAHFqUmZ+AApgAE82ahYAMxjTeOBqAEpsgBpsEPDI6L44xOS0vn0mXIKiiKimWISk1Ji8TFyAXX8fOAc8ZwpybC4AR2JIgTkFA2oAFRYAHgWAPjjdCggxtjBSVbXsCD3SAsJqMAATAHlSJXiOU8PsFmBdZmy+3x+fAHcPkxqNgFtgQNg6LM9kwzCpDt8vGhfr4KClsOU2mkFtlsABeXHYVqVdKZHJfJHI35wADsY2ok2m1Ch8yWMW+lN+QLRAFVSBAwClaABBCCrAoZLIbLnYACy1CgLCY8RlBgg+iU+AiwAugqBpAEMRO+zq7I5viNpHwRkiAHN3nlTWbsNLefzBfgRWLsBLqFLqGi5QqlSqmGqwBqtTrmIyYm9ASaKU6fnHmFbbe9sjZHdhUeiie1sXiCfmqjVctmfjS6Qz9Uz5NDFiw2Ymkz4XXyBcLRQsCtUDH6A/LFcrVerNdRtf7o7XDacE62fha06Q7boHS3W+23V2vX2mAPZUPg6Pw+PJ7qYynavkK5Sr8vV5n0NmqxMprXmUDWdLA8OQ2GIwnKM9QNC18mOU4H3tZ1/UPIMR1DMdIynEDqFjd5PgKe9rRXDMbB+ABfZ8WwCOV3hYK46AUJQYjI3QKKotglBuNhKBYPl8AAORYBglBGEYbWyXoWzYZpuDALInBzUY30ZT9GxiaUfW9JAChGYBsCUVSYLRZTTAQNTRiUBByR+REOSrTB8N8IiW2+UiJ3oyjqNoxyGOoli2I47jeP4lchO+USKHEyShmkjTZI/esWSbaU9xUwyNK0go4pqb0DPCzSTIRF9aSs75bJ+ey6PcpjXPI5ymM8ih2IgLiePuPzBOEn4gpC4EwvU6t3xmaKvxWdZFNg0FTG0rrkp0kF0sSrLTN8czKUs6yfFsgigA==)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs [hunk 4]
```diff
@@ -16,6 +16,8 @@ public enum JsonNumberHandling
         /// <summary>
         /// Numbers can be read from <see cref="JsonTokenType.String"/> tokens.
         /// Does not prevent numbers from being read from <see cref="JsonTokenType.Number"/> token.
+        /// Strings that have escaped characters will be unescaped before reading. Leading or trailing trivia,
+        /// including whitespace, is not allowed.
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs:20]:
```suggestion
        /// Strings that have escaped characters will be unescaped before reading. Leading or trailing trivia within the string token,
        /// including whitespace, is not allowed.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs [hunk 5]
```diff
@@ -25,7 +27,10 @@ public enum JsonNumberHandling
         /// <summary>
         /// The "NaN", "Infinity", and "-Infinity" <see cref="JsonTokenType.String"/> tokens can be read as
         /// floating-point constants, and the <see cref="float"/> and <see cref="double"/> values for these
-        /// constants will be written as their corresponding JSON string representations.
+        /// constants (such as <see cref="float.PositiveInfinity"/> and <see cref="double.NaN"/>)
+        /// will be written as their corresponding JSON string representations.
+        /// Strings that have escaped characters will be unescaped before reading.
+        /// Leading or trailing trivia, including whitespace, is not allowed.
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs:33]: So, with such a setting, "\u0031" will get parsed as the integer number 1. Neat.

File: src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs [hunk 6]
```diff
@@ -814,19 +808,27 @@ static void PerformFloatingPointSerialization(string testString)
...
         [InlineData(" Infinity")]
         [InlineData(" -Infinity")]
-        [InlineData("NaN ")]
         [InlineData("Infinity ")]
         [InlineData("-Infinity ")]
         [InlineData("a-Infinity")]
         [InlineData("NaNa")]
         [InlineData("Infinitya")]
         [InlineData("-Infinitya")]
+        [InlineData("\u006EaN")] // "naN"
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs:820]: Nice test around escaped casing.

**Review by ahsonkhan — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberHandling.cs:33]: Yes, when `AllowReadingFromString` is active (these comments are for `AllowNamedFloatingPointLiterals`).

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:231]: It came from saving this file in VS after deleting the `JsonSerializer.cs` file with the GUI (also in VS).

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs]: Local benchmarks show using `(x == typeof(...))` can be faster and has potentially lower code size than using a cached `Type` instance. I made a minor change in this PR to align better with this, but plan to do an audit of how we use `typeof()` in a follow up PR backed with microbenchmarks/JSON benchmarks.

**Review by ahsonkhan — APPROVED** [review]

**Review by steveharter — APPROVED** [review]

**Comment by steveharter** [comment]
@layomia note there was a STJ failure in `(Libraries Test Run release mono Windows_NT x64 Debug`. Appeared the tests didn't run or finish since I didn't see any output. I'll try to restart.

**Comment by layomia** [comment]
> note there was a STJ failure in (Libraries Test Run release mono Windows_NT x64 Debug

Thanks. It passed in a [subsequent run](https://dev.azure.com/dnceng/public/_build/results?buildId=827688&view=logs&jobId=d5c01a48-52b8-51d9-fe3a-6804ba4b63f8). I don't think any issue there is due to this PR. I'll watch out for failures in other PRs.

---

Test failures appear unrelated - 3 timeouts:

`runtime (Build Browser wasm Release AllSubsets_Mono) Cancelled after 120m`

`runtime (Libraries Test Run release coreclr Linux x64 Debug) Cancelled after 150m`

`runtime (Libraries Test Run release mono Linux x64 Debug) Cancelled after 150m`
